# FishFinder
### Sam Virsik

Just a note, for some reason my first GitHub repo I set up for this did not work. I kept pushing to it every class and it seemed to my like it was working, but I checked when the project was just about over and none of the code had gone through. I set up a new repo and tried to chunk out the pushes to GitHub as I think I tried to do them as the project went along, but it won't be perfect. Sorry about that. 

In [3]:
import tifffile as tiff
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
from PIL import Image
#import random
import time
import imagecodecs
#import scipy.sparse as sp
import matplotlib.cm as cm

C:\Users\SamV\AppData\Local\Temp\ipykernel_19444\129697075.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


In [4]:
#A function to get the data from the tif/tiff file and convert it to a dataframe
def readData(importName):
    tif_data = tiff.imread(importName)
    df = pd.DataFrame(tif_data)
    number_rows, number_columns = df.shape
    return df

In [5]:
#Getting the data
img_1 = readData("data\img_1_data.tif")
img_4 = readData("data\img_4_data.tif")
img_7 = readData("data\img_7_data.tif")
img_9 = readData("data\img_9_data.tif")
img_20 = readData("data\img_20_data.tiff") #This is a tiff, not a tif, since for some reason they decided to change the file format halfway through my project. It's fine, the code works identically on it. 

## Data Exploration 

In [6]:
img_4.head()

0         1         2         3         4         5         6    \
0 -3.237103 -3.127103 -3.317103 -2.892198 -2.072198 -2.262198 -2.802198   
1 -3.447104 -3.717103 -3.327103 -2.792198 -2.962198 -3.752198 -4.412198   
2 -3.623092 -4.083092 -3.303092 -3.315799 -3.905799 -4.455800 -5.125800   
3 -4.443092 -4.363092 -4.963092 -4.445799 -3.625799 -4.215799 -3.645799   
4 -5.423092 -6.053092 -4.473093 -4.225800 -4.455800 -4.195799 -3.545799   

        7         8         9    ...       266       267       268       269  \
0 -3.002198 -3.597368 -4.177368  ... -8.609853 -8.799852 -8.989853 -9.069853   
1 -4.512198 -3.887367 -3.907367  ... -8.419852 -8.199852 -8.209852 -8.639853   
2 -3.725800 -3.737474 -4.157475  ... -8.539809 -8.279809 -8.279809 -8.449809   
3 -3.285800 -3.647475 -3.317475  ... -9.459809 -9.179810 -9.019810 -8.919809   
4 -3.245800 -3.057475 -2.767474  ... -9.269810 -9.159809 -9.169809 -8.989809   

        270       271       272       273       274       275  
0 -8.741062 -8.551063 -8.501062 -8.651062 -8.731063 -8.395384  
1 -8.931063 -8.801063 -8.611062 -8.211062 -8.041062 -7.525384  
2 -8.853113 -8.923113 -8.813112 -8.803113 -8.283113 -7.705556  
3 -8.983112 -8.853113 -8.863112 -8.613112 -8.463113 -8.115556  
4 -8.963113 -8.953113 -8.843113 -8.393113 -7.873113 -7.645556  

[5 rows x 276 columns]

In [7]:
img_4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107 entries, 0 to 106
Columns: 276 entries, 0 to 275
dtypes: float32(276)
memory usage: 115.5 KB


In [8]:
img_4.describe()

0           1           2           3           4           5    \
count  107.000000  107.000000  107.000000  107.000000  107.000000  107.000000   
mean    -8.085320   -8.047469   -7.797095   -7.662092   -7.537138   -7.526298   
std      1.930470    1.909606    2.036478    2.157805    2.348795    2.406461   
min    -12.343512  -12.273511  -12.253511  -12.233159  -12.223159  -12.293159   
25%     -9.353034   -9.203026   -9.296703   -9.342427   -9.326018   -9.249531   
50%     -7.845592   -7.717869   -7.425592   -7.469675   -7.204831   -7.069675   
75%     -6.698570   -6.698936   -6.287190   -6.141601   -6.121601   -6.022331   
max     -3.237103   -3.127103   -3.303092   -2.792198   -2.072198   -2.262198   

              6           7           8           9    ...         266  \
count  107.000000  107.000000  107.000000  107.000000  ...  107.000000   
mean    -7.494522   -7.392092   -7.273218   -7.207143  ...  -60.861210   
std      2.443609    2.595504    2.685459    2.790987  ...   38.868069   
min    -12.333158  -12.353159  -12.778747  -12.978746  ... -131.714676   
25%     -9.199530   -8.942427   -8.818319   -8.822290  ...  -92.094910   
50%     -6.949675   -6.989675   -7.382732   -7.456027  ...  -62.866013   
75%     -5.892331   -5.709818   -5.291938   -5.077312  ...  -24.159143   
max     -2.802198   -2.888912   -2.487461   -2.397461  ...   -8.183590   

              267         268         269         270         271         272  \
count  107.000000  107.000000  107.000000  107.000000  107.000000  107.000000   
mean   -61.335224  -61.571308  -62.008781  -62.401749  -62.645584  -63.236141   
std     39.086292   39.185768   39.353317   39.556744   39.629410   39.743202   
min   -132.434677 -132.774689 -133.444687 -134.109726 -134.449738 -135.129730   
25%    -93.139904  -93.774910  -94.929909  -96.082996  -96.587994  -97.472996   
50%    -63.056011  -63.156010  -63.536011  -64.077629  -64.397629  -65.057625   
75%    -24.239143  -24.219143  -24.119143  -23.963130  -23.928129  -24.183130   
max     -7.923590   -7.773590   -7.723590   -7.905536   -8.040492   -8.123113   

              273         274         275  
count  107.000000  107.000000  107.000000  
mean   -63.734276  -63.931377  -64.370827  
std     39.945080   40.085857   40.280655  
min   -135.819733 -136.169739 -136.858200  
25%    -98.127998  -98.412998  -98.907494  
50%    -65.667625  -65.927628  -66.413010  
75%    -24.333130  -24.603129  -25.740531  
max     -7.470492   -7.473113   -7.282321  

[8 rows x 276 columns]

In [9]:
img_4.isnull().sum()

0      0
1      0
2      0
3      0
4      0
      ..
271    0
272    0
273    0
274    0
275    0
Length: 276, dtype: int64

From this basic data exporation we see that we have a 276x107 dataframe of depth values. The negative values are for the ocean, and the positive values are for land. This kind of data exploration was only shown on image 4 (because it is of reasonable size), but I also did it on all other images used to make sure we had no large irregularities in the data. 

## Data Cleaning

In [10]:
img_20

0          1          2          3          4          5    \
0   -11.348689 -11.348689 -11.328690 -11.328690 -11.328690 -11.328690   
1   -11.348689 -11.348689 -11.328690 -11.328690 -11.328690 -11.328690   
2   -11.288690 -11.288690 -11.278689 -11.278689 -11.278689 -11.288690   
3   -11.288690 -11.288690 -11.278689 -11.278689 -11.278689 -11.288690   
4   -11.288690 -11.288690 -11.278689 -11.278689 -11.278689 -11.288690   
..         ...        ...        ...        ...        ...        ...   
389  -8.494452  -8.494452  -8.444451  -8.444451  -8.444451  -8.364451   
390  -8.494452  -8.494452  -8.444451  -8.444451  -8.444451  -8.364451   
391  -8.494452  -8.494452  -8.444451  -8.444451  -8.444451  -8.364451   
392  -8.154451  -8.154451  -8.144452  -8.144452  -8.144452  -8.114451   
393  -8.154451  -8.154451  -8.144452  -8.144452  -8.144452  -8.114451   

           6          7          8          9    ...        757        758  \
0   -11.328690 -11.328690 -11.358689 -11.358689  ...  -5.182217  -5.182217   
1   -11.328690 -11.328690 -11.358689 -11.358689  ...  -5.182217  -5.182217   
2   -11.288690 -11.288690 -11.308689 -11.308689  ...  -5.512217  -5.512217   
3   -11.288690 -11.288690 -11.308689 -11.308689  ...  -5.512217  -5.512217   
4   -11.288690 -11.288690 -11.308689 -11.308689  ...  -5.512217  -5.512217   
..         ...        ...        ...        ...  ...        ...        ...   
389  -8.364451  -8.364451  -8.264452  -8.264452  ... -14.912817 -14.912817   
390  -8.364451  -8.364451  -8.264452  -8.264452  ... -14.912817 -14.912817   
391  -8.364451  -8.364451  -8.264452  -8.264452  ... -14.912817 -14.912817   
392  -8.114451  -8.114451  -8.034451  -8.034451  ... -15.242817 -15.242817   
393  -8.114451  -8.114451  -8.034451  -8.034451  ... -15.242817 -15.242817   

           759        760        761        762        763        764  \
0    -5.182217  -5.562217  -5.562217  -5.562217  -6.142217  -6.142217   
1    -5.182217  -5.562217  -5.562217  -5.562217  -6.142217  -6.142217   
2    -5.512217  -5.892217  -5.892217  -5.892217  -6.272216  -6.272216   
3    -5.512217  -5.892217  -5.892217  -5.892217  -6.272216  -6.272216   
4    -5.512217  -5.892217  -5.892217  -5.892217  -6.272216  -6.272216   
..         ...        ...        ...        ...        ...        ...   
389 -14.912817 -14.542817 -14.542817 -14.542817 -14.172817 -14.172817   
390 -14.912817 -14.542817 -14.542817 -14.542817 -14.172817 -14.172817   
391 -14.912817 -14.542817 -14.542817 -14.542817 -14.172817 -14.172817   
392 -15.242817 -15.142817 -15.142817 -15.142817 -15.222817 -15.222817   
393 -15.242817 -15.142817 -15.142817 -15.142817 -15.222817 -15.222817   

           765        766  
0    -6.142217  -6.602217  
1    -6.142217  -6.602217  
2    -6.272216  -6.292216  
3    -6.272216  -6.292216  
4    -6.272216  -6.292216  
..         ...        ...  
389 -14.172817 -14.752817  
390 -14.172817 -14.752817  
391 -14.172817 -14.752817  
392 -15.222817 -16.072817  
393 -15.222817 -16.072817  

[394 rows x 767 columns]

If you look closely, you can see that many of the values in that dataframe are repeats. If this were truly good satellite data, it would be near impossible for that to happen. This tells us that we likely have some kind of error in our dataset. 

Let's take a closer look at how we can clean this data.

In [11]:
num_duplicates = img_20.duplicated().sum()

percent = num_duplicates / img_20.shape[0]

print("Percent of rows with duplicates: " + "{:.1%}".format(percent))

Percent of rows with duplicates: 66.5%


Clearly, with 66.5% of values being duplicates, there is a big issue we need to address before we can effectively analyze this data. 

What I noticed however, is that there seems to be a pattern to how the data is repeated: There are entire columns and entire rows which are repeats (not just random singular values). 

My suspicion is that the satellite data is not actually 1/9 resolution, and is more like 1/3 or even 1 arcsecond. Then to conform to the NOAA data viewer they simply duplicated any unknown higher resolutions. 

By simply cutting out the repeated rows and columns we should be able to reduce the data to the highest ACTUAL resolution we have. We won't be sure what this resolution is, but we know it will not overrepresent the granularity of the data. 

In [12]:
#Function to remove the duplicated values that we saw in img_20. It removes duplicated rows, then transposes and 
#does that again, then reindexs so that the dataframe is not 'highly fragmented' 

def remove_repeats(img_20):
    img_20 = img_20.drop_duplicates()
    img_20 = img_20.transpose()

    img_20 = img_20.drop_duplicates()
    img_20 = img_20.transpose()

    max_row = len(img_20.index)
    max_col = len(img_20.columns)
    img_20.columns = range(max_col)
    img_20.index = range(max_row)

    return(img_20)

I simply removed any duplicated rows from the data, then transposed it and removed any dupliated columns. I then transposed it again and reindexed it so it wasn't fragmented and strangely formatted. 

For most datasets this would not be an operation performed. However, since the columns of my data are not 'categories' and my rows are not 'items,' this kind of manipulation actually makes a lot of sense. 

In [13]:
img_20 = remove_repeats(img_20)
img_20

0          1          2          3          4          5    \
0   -11.348689 -11.328690 -11.328690 -11.358689 -11.398689 -11.442095   
1   -11.288690 -11.278689 -11.288690 -11.308689 -11.338690 -11.382095   
2   -11.218689 -11.198689 -11.208690 -11.238689 -11.268689 -11.302095   
3   -11.128690 -11.118690 -11.128690 -11.158689 -11.188689 -11.222095   
4   -11.045706 -11.035706 -11.055706 -11.075706 -11.115705 -11.152165   
..         ...        ...        ...        ...        ...        ...   
127  -9.464452  -9.304452  -9.104452  -8.884452  -8.654451  -8.422848   
128  -9.174452  -9.044452  -8.884452  -8.694451  -8.504452  -8.302848   
129  -8.854452  -8.754452  -8.634452  -8.484451  -8.334452  -8.182848   
130  -8.494452  -8.444451  -8.364451  -8.264452  -8.154451  -8.032847   
131  -8.154451  -8.144452  -8.114451  -8.034451  -7.954452  -7.862848   

           6          7          8          9    ...        247        248  \
0   -11.502095 -11.542095 -11.552095 -11.542095  ...  -7.774415  -8.014415   
1   -11.422095 -11.452095 -11.472095 -11.472095  ...  -7.004415  -6.934415   
2   -11.342095 -11.372095 -11.392095 -11.402095  ...  -6.154415  -5.834414   
3   -11.262095 -11.292095 -11.312095 -11.332095  ...  -5.304415  -4.814415   
4   -11.182165 -11.212165 -11.242166 -11.262165  ...  -4.598418  -3.988418   
..         ...        ...        ...        ...  ...        ...        ...   
127  -8.202847  -8.012848  -7.992848  -8.102848  ... -16.009518 -15.849517   
128  -8.122848  -7.992848  -7.962848  -8.022848  ... -14.209517 -14.109517   
129  -8.042848  -7.952848  -7.922848  -7.942848  ... -15.119516 -14.609517   
130  -7.932848  -7.862848  -7.842848  -7.852848  ... -15.699516 -15.529517   
131  -7.782848  -7.732848  -7.722848  -7.742848  ... -14.609517 -14.889517   

           249        250        251        252        253        254  \
0    -8.014415  -7.144415  -6.174415  -5.434415  -5.182217  -5.562217   
1    -6.714415  -6.184415  -5.704415  -5.434415  -5.512217  -5.892217   
2    -5.464415  -5.084414  -5.004415  -5.174415  -5.572217  -6.172216   
3    -4.364415  -4.004415  -4.244415  -4.704415  -5.342216  -6.172216   
4    -3.618418  -3.448418  -3.658418  -4.118418  -4.728772  -5.398771   
..         ...        ...        ...        ...        ...        ...   
127 -17.399517 -16.929516 -16.589518 -15.609517 -15.522817 -16.032818   
128 -14.309517 -14.839517 -15.479517 -15.209517 -14.852818 -15.242817   
129 -14.579516 -13.939517 -14.689517 -14.929517 -14.722817 -14.592817   
130 -16.409517 -15.669517 -15.139517 -15.189517 -14.912817 -14.542817   
131 -15.159516 -14.939517 -14.609517 -15.009517 -15.242817 -15.142817   

           255        256  
0    -6.142217  -6.602217  
1    -6.272216  -6.292216  
2    -6.632216  -6.412217  
3    -6.942216  -6.442216  
4    -5.858771  -5.638772  
..         ...        ...  
127 -16.892817 -16.162817  
128 -15.532817 -15.022817  
129 -14.332817 -14.082817  
130 -14.172817 -14.752817  
131 -15.222817 -16.072817  

[132 rows x 257 columns]

This data looks much better - the repeats seem to be gone. Let's give it the same test for repeats: 

In [14]:
num_duplicates = img_20.duplicated().sum()

percent = num_duplicates / img_20.shape[0]

print("Percent of rows with duplicates: " + "{:.1%}".format(percent))

Percent of rows with duplicates: 0.0%


## Exploratory Data Analysis

My datset is not typical in that the rows and columns do not define categories or instances of data. They actually are meant to organize the data by GPS coordinates. Thus, my exploratory data analysis will be a bit different from other datasets. I will process the data and turn it into various images so that I can better visualize it. 

In [15]:
def analysis_1(df, width): #5 foot

    number_rows, number_columns = df.shape
    image = Image.new('RGB', (number_columns, number_rows), color = 'white')
    pixels = image.load()

    rowOnePercent = number_rows/100
    currentRow = 0
    rowNeededToChange = rowOnePercent
    nextPercent = 1
    print("")
    print("--------------------------")
    
    colorList = [(139, 69, 19), (212, 255, 1), (190, 255, 1), (166, 255, 17), (139, 255, 17), 
                 (115, 255, 49), (83, 255, 84), (53, 255, 114), (53, 255, 147), (53, 255, 199), 
                 (17, 255, 220), (0, 255, 249), (0, 245, 249), (0, 226, 249), (0, 212, 242), 
                 (0, 198, 242), (0, 182, 242), (0, 167, 242), (0, 150, 242), (0, 128, 242), 
                 (0, 110, 218), (0, 88, 218), (0, 61, 215), (0, 30, 215), (0, 0, 207), (0, 0, 187), 
                 (0, 0, 163), (0, 0, 140), (0, 0, 128), (0, 0, 108), (0, 0, 88)]

    for row_index, row_data in df.iterrows():
        for col_name in df.columns:
            mValue = row_data[col_name]
            fValue = mValue*3.28084 #foot value

            scaled = ((abs(fValue)) - fValue) / 2
            
            if scaled == 0.0:
                pixels[col_name, row_index] = colorList[0]
            else:
                scaled = int(scaled/width) + 1
            
            if (scaled > len(colorList)-2):
                pixels[col_name, row_index] = colorList[len(colorList)-1]
            else:
                pixels[col_name, row_index] = colorList[int(scaled)]

        if(row_index > rowNeededToChange):
            printString = "Progress: " + str(nextPercent) + "%"
            print('\r' + printString, end='')
            rowNeededToChange = rowNeededToChange + rowOnePercent
            nextPercent +=1

    finalPrint = "Progress: 100%"
    print('\r' + finalPrint)
    
    return image

This is the first anaysis function. It simply creates a variably chunked heatmap of the image, using predefined colors. I have found this analysis to be good for generally getting a sense of the image. It also doesn't take too long to run, since I changed the way I do it to be a function instead of a list of conditionals. There is a progress bar to make it more user friendly when running analysis on larger images. Let's see how it looks on some of our images. 

In [16]:
myImage4 = analysis_1(img_4, 2)
myImage4.show()


--------------------------
Progress: 100%


Image 4 is very low resolution, and is only used to test computation and functions before running them on larger images. Since it looks good we can continue to bigger images. 

In [17]:
myImage9 = analysis_1(img_9, 0.5)
myImage9.show()


--------------------------
Progress: 100%


Image 9 looks good as well. This is an image of the water off of stirrup key. I go fishing here a lot. 

In [ ]:
myImage7 = analysis_1(img_7, 5)
myImage7.show()

This is a much more useful image. This captures much of the water on the ocean side of the florida keys, and is of mnoderately high resolution. 

I actually went through the entire image looking for irregularities in the data, which could be good fishing spots. I found one interesting spot a bit further out than the reef about below stirrup key. I know that will not mean anything to most people, so I grabbed the spot again in even higher resolution data. Let's take a look. 

In [ ]:
myImage1 = analysis_1(img_1, 5)
myImage1.show()

It was hard to grab data for only the exact area that I wanted, so it is still hard to see exactly what I mean. However, take a look at the top left corner of that image in the medium dark blue water. You will notice a lighter area. 

That is actually a shipwreck - the Thunderbolt - and is a great spot for fishing. I have SCUBA dived there before, and the fishing on the wreck is great. 

In [18]:
myImage20 = analysis_1(img_20, 0.35)
myImage20.show()


--------------------------
Progress: 100%


This is all interesting, but there is only so much value that one is able to extract from manually looking at a heatmap of an image. Let's take a look at another algorithm, one which makes a black and white image of depth change. Since fish like structure, the places with the greatest depth change will often be the best for fishing. 

In [19]:
def analysis_2(df):

    df_new = pd.DataFrame(False, index=df.index, columns=df.columns)

    num_rows, num_cols = df.shape
    image = Image.new('RGB', (num_cols, num_rows), color = 'white')
    pixels = image.load()


    for i in range(df.shape[0] - 1, 0, -1):
        for col in df.columns:
            try:
                if df.at[i, col] >= 0:
                    pixels[col, i] = (139, 69, 19)
                elif df.at[i, col] > df.at[i - 1, col]:
                    pixels[col, i] = (0, 0, 0)
            except:
                pass

    return image

This function simply makes pixels black if there was an increase in depth from the pixel above and white if not. When looking close up at an image it will be hard to tell what is going on, but when you zoom out it becomes easier to see the ridges and ledges than in the heatmap. 

In [21]:
img_4 = readData("data\img_4_data.tif")
myImage4 = analysis_2(img_4)
myImage4.show()

Now that we know image 4 works, we can move onto large images.

In [ ]:
img_9 = readData("data\img_9_data.tif")
myImage9 = analysis_2(img_9)
myImage9.show()

In [ ]:
img_7 = readData("data\img_7_data.tif")
myImage7 = analysis_2(img_7)
myImage7.show()

In [ ]:
img_20 = readData("data\img_20_data.tiff")
img_20 = remove_repeats(img_20)
myImage20 = analysis_2(img_20)
myImage20.show()

Let's take a look at amnother type of function to explore the data. 

In [23]:
#Most of the code in this block was written by GPT 3.5, and edited by myself

def number_to_color(value, vmin, vmax, cmap):
    norm_value = (value - vmin) / (vmax - vmin)
    rgba_color = cmap(norm_value)
    r, g, b, _ = [int(255 * x) for x in rgba_color]
    return r, g, b

def analysis_3(df):
    vmin = df.min().min()
    vmax = df.max().max()


    width, height = df.shape[1], df.shape[0]
    image_data = []
    cmap = cm.get_cmap('plasma') #or cool, or plasma, or inferno, magma, jet, hot, gray, cividis

    for row in df.values:
        for value in row:
            r, g, b = number_to_color(value, vmin, vmax, cmap)
            image_data.append((r, g, b))

    image = Image.new('RGB', (width, height))
    image.putdata(image_data)

    return image

This is a matplotlib type color map. I personally don't like it as much as my color map, since it is harder to see separation between various depth ranges. However it is technically more accurate at putting the pixels into their associated categories since there are a lot more colors used.

In [24]:
img_4 = readData("data\img_4_data.tif")
myImage4 = analysis_3(img_4)
myImage4.show()

C:\Users\SamV\AppData\Local\Temp\ipykernel_19444\2252383214.py:16: MatplotlibDeprecationWarning: The get_cmap function was deprecated in Matplotlib 3.7 and will be removed two minor releases later. Use ``matplotlib.colormaps[name]`` or ``matplotlib.colormaps.get_cmap(obj)`` instead.
  cmap = cm.get_cmap('plasma') #or cool, or plasma, or inferno, magma, jet, hot, gray, cividis


In [ ]:
img_7 = readData("data\img_7_data.tif")
myImage7 = analysis_3(img_7)
myImage7.show()

In [ ]:
img_20 = readData("data\img_20_data.tiff")
img_20 = remove_repeats(img_20)
myImage20 = analysis_3(img_20)
myImage20.show()

## Statistical Data Analysis

Since my data does not have clear categories in the same way, my statistical analysis is different from that of most datasets. I have aimed to statistically analyze my entire datasets at once, since column or row based analysis doesn't make sense in this context. 

I have thought of a couple of statistics I would be interested in knowing: 

##### Mean and standard deviation of depth values

##### Mean and standard deviation of depth change (defined likely as change in some predefined direction)

##### Greatest depth change

##### Percent of spots with a depth change greater than some value

I will run most of these analysis on img 9, since it is of fair size and quality. 

I might also run some on image 7, since that captures oceanside data which is more interesting. 

### Mean and standard deviation of depth values:


In [26]:
img_9 = readData("data\img_9_data.tif")

mean_value = img_9.values.mean() #remember, this is in METERS
std_value = img_9.values.std()

print("Mean: " + str(mean_value))
print("Standard Deviation: " + str(std_value))

Mean: -2.799797
Standard Deviation: 0.6427821


This gives us a better sense of the general depth of the image. Remeber that -2 means 2 meters below the surface. This is actually an interesting statistic, since the heatmap is unclear about just how deep into the next color range those values go. This helps clear that up. 

### Mean and standard deviation of depth change

In [ ]:
img_9 = readData("data\img_9_data.tif")

df_new = pd.DataFrame(index=img_9.index, columns=img_9.columns)

num_rows, num_cols = img_9.shape

for row in range(img_9.shape[0] - 1, 0, -1):
    for col in img_9.columns:
        df_new.iloc[row, col] = img_9.iloc[row, col] - img_9.iloc[row-1, col]

df_new.iloc[0] = 0

print("Mean: " + str(df_new.values.mean()))
print("Standard Deviation: " + str(df_new.values.std()))

In [ ]:
img_7 = readData("data\img_7_data.tif")

df_new = pd.DataFrame(index=img_7.index, columns=img_7.columns)

num_rows, num_cols = img_7.shape

for row in range(img_7.shape[0] - 1, 0, -1):
    for col in img_7.columns:
        df_new.iloc[row, col] = img_7.iloc[row, col] - img_7.iloc[row-1, col]

df_new.iloc[0] = 0

print("Mean: " + str(df_new.values.mean()))
print("Standard Deviation: " + str(df_new.values.std()))

### Greatest depth change

In [ ]:
df_new.max().max()

This is pretty crazy! The greatest depth change in a single pixel (in this resolution about 30 feet) is around 20 feet. This would be a good spot for fishing, since there is likely a lot of holes and structure for the fish to congregate on. 

### Percent of spots with a depth change greater than some value

In [ ]:
count_greater_than_2 = (df_new > 2).sum().sum()

total_values = df_new.size

percentage_greater_than_2 = (count_greater_than_2 / total_values) * 100

percentage_greater_than_2

This is interesting. The greatest depth change was 18 feet, but there are less than 0.021 percent of data points which ever even make it to 6 feet change. This means that our max depth change value is even more interesting, and is definately a spot worth checking out. 

Thank you for checking out FishFinder! This is only some of the algorithms and images I have done analysis on, but it has been a very interesting project for me so far. I plan on continuing it throughout the year. 